etcd/v3/kv_etcd.go

```diff
@@ -1497,53 +1497,84 @@ func (et *etcdKV) RemoveMember(
 }
 
 func (et *etcdKV) ListMembers() (map[string]*kvdb.MemberInfo, error) {
+	var (
+		fnMemberStatus = func(cliURL string) (*kvdb.MemberInfo, uint64, error) {
+			if cliURL == "" {
+				return nil, 0, fmt.Errorf("Must provide client URL")
+			}
+			// Use the context with no leader requirement as we might be hitting
+			// an endpoint which is down
+			ctx, cancel := et.MaintenanceContext()
+			endpointStatus, err := et.maintenanceClient.Status(ctx, cliURL)
+			cancel()
+
+			if err != nil {
+				return nil, 0, err
+			}
+
+			mid := endpointStatus.Header.MemberId
+			return &kvdb.MemberInfo{
+				// PeerUrls:   .. must fill later
+				ClientUrls: []string{cliURL},
+				Leader:     endpointStatus.Leader == mid,
+				DbSize:     endpointStatus.DbSize,
+				IsHealthy:  true,
+				ID:         strconv.FormatUint(mid, 16),
+			}, mid, nil
+		}
+	)
 	ctx, cancel := et.MaintenanceContextWithLeader()
 	memberListResponse, err := et.kvClient.MemberList(ctx)
 	cancel()
+
 	if err != nil {
 		return nil, err
 	}
-	resp := make(map[string]*kvdb.MemberInfo)
+
+	membersMap := make(map[uint64]*kvdb.MemberInfo)
 	mLock.Lock()
 	defer mLock.Unlock()
+
+	// Get status from Endpoints
+	for _, ep := range et.GetEndpoints() {
+		mi, mid, err := fnMemberStatus(ep)
+		if err != nil || mi == nil {
+			logrus.WithError(err).Warnf("kvClient.Status(%s) returned error", ep)
+			continue
+		}
+		membersMap[mid] = mi
+	}
+
+	// Get status from MemberList() nodes, that were missing from GetEndpoints() list
 	for _, member := range memberListResponse.Members {
-		var (
-			leader     bool
-			dbSize     int64
-			isHealthy  bool
-			clientURLs []string
-		)
-		// etcd versions < v3.2.15 will return empty ClientURLs if
-		// the node is unhealthy. For versions >= v3.2.15 they populate
-		// ClientURLs but return an error status
-		if len(member.ClientURLs) != 0 {
-			// Use the context with no leader requirement as we might be hitting
-			// an endpoint which is down
-			ctx, cancel := et.MaintenanceContext()
-			endpointStatus, err := et.maintenanceClient.Status(
-				ctx,
-				member.ClientURLs[0],
-			)
-			cancel()
-			if err == nil {
-				if member.ID == endpointStatus.Leader {
-					leader = true
+		for _, cu := range member.ClientURLs {
+			if _, has := membersMap[member.ID]; !has {
+
+				mi, mid, err := fnMemberStatus(cu)
+				if err != nil || mi == nil {
+					logrus.WithError(err).Warnf("kvClient.Status(%s) returned error", cu)
+					continue
 				}
-				dbSize = endpointStatus.DbSize
-				isHealthy = true
-				// Only set the urls if status is healthy
-				clientURLs = member.ClientURLs
+				membersMap[mid] = mi
 			}
 		}
-		resp[member.Name] = &kvdb.MemberInfo{
-			PeerUrls:   member.PeerURLs,
-			ClientUrls: clientURLs,
-			Leader:     leader,
-			DbSize:     dbSize,
-			IsHealthy:  isHealthy,
-			ID:         strconv.FormatUint(member.ID, 16),
+	}
+
+	// Fill PeerURLs; also, remap with "Name" as a key
+	resp := make(map[string]*kvdb.MemberInfo)
```

Comment:
Why not start with this map itself and populate it? `member.Name` is also a unique key. So essentially you first loop over the endpoints and fill the `resp`. And then loop over the member.List() output and populate the remaining entries in this `resp`.  This would avoid an extra map and propbably reduce a few loops?

Reply:
Oh.. didn't want to change the return-type of the function -- it's still `map[string]*MemberInfo` as before, rather than `map[int64]*MemberInfo` that we have to use for `unique(<byID>)` effect.

As for the loops - I didn't want to super-optimize it, since the code would be less readable, also 99% of the time we'll be dealing with 3-elements "in-memory" "maps of pointers"  (= fast access, not much CPU lost).

I'm more worried about the fact that the whole `ListMembers()` func is actually called every few seconds, and we're potentially significantly slowing it down when "KVDB members are missing"  (e.g. 7 sec timeout on `Stat()` via `GetEndpoints()` URL, then _another_ 7 sec timeout on `Stat()` via `MemberList()` URL)

Reply:
@adityadani ?  Are you OK with these comments, or you're asking me to change the code?

Reply:
> But I still don't get the your unique<byID> comment

The `member.ID` is actually `int64` type -- so while looking for unique members, we will rely on a `map[int64]` structure.
The structure we need to return is keyed to `member.Name`, which is a `string`-type.  This is the reason why we "build two maps" -- but note that both maps have mostly 3 elements inside (+ pointer to struct), so they won't take much space.

> Can we move the for loop over the MemberList output before the for loop that goes over `GetEndpoints()` list.

I thought the `GetEndpoints()` would be more reliable, i.e.
* `GenEndpoints()` is the list of endpoints explicitly given to client as KVDB-configuration, while
* `MemberList()` is the list of EtcD's internal peers, which may (or may not) be accessible to the client

So, it made sense to go through the `GenEndpoints()` first, as there is higher probability all the endpoints given to the client will work -- and then "fill up the gaps" by going through the "less reliable" `MemberList()` list of endpoints.

Please let me know if you still think this is not OK, and I'll change the code.